File: ocaml/2015/day03a/day03a.ml
open Printf
open Sys

module Position2D =
  struct
    type t = int * int
    let compare (x0, y0) (x1, y1) =
      match Stdlib.compare x0 x1 with
      | 0 -> Stdlib.compare y0 y1
      | res -> res
  end

module PositionSet = Set.Make(Position2D)

let usage progname =
  eprintf "usage: %s <input file>\n" progname;
  exit 1

let process_file filename =
  let process_ch (santa, positions) ch = (
    let (santa_x, santa_y) = santa in
    let new_santa = ( match ch with
                      | '^' -> (santa_x, santa_y + 1)
                      | 'v' -> (santa_x, santa_y - 1)
                      | '<' -> (santa_x - 1, santa_y)
                      | '>' -> (santa_x + 1, santa_y)
                      | _ -> santa ) in
    let new_positions = PositionSet.add new_santa positions in
    (new_santa, new_positions)
  )
  in
  try
    let content = In_channel.input_all (In_channel.open_text filename) in
    let (_, final_positions) = String.fold_left process_ch ((0, 0), PositionSet.empty) content in
    PositionSet.cardinal final_positions
  with
  | Sys_error msg ->
      eprintf "error reading file: %s\n" msg;
      exit 1

let () =
  let args = Sys.argv
  in if Array.length args < 2 then
       usage args.(0) ;
     let filename = args.(1) in
     let result = process_file filename in
     printf "result = %d\n" result

File: nim/2015/day03a/day03a.nim
import os, sets

type
  Position2D = object
    x, y: int

proc `==`(a, b: Position2D): bool =
  a.x == b.x and a.y == b.y

proc hash(a: Position2D): int =
  (a.x shl 16) or a.y

proc processFile(filename: string): int =
  var positions = initHashSet[Position2D]()
  var santa = Position2D(x: 0, y: 0)
  positions.incl(santa)
  for line in lines(filename):
    for ch in line:
      if ch == '^': santa = Position2D(x: santa.x, y: santa.y + 1)
      elif ch == 'v': santa = Position2D(x: santa.x, y: santa.y - 1)
      elif ch == '<': santa = Position2D(x: santa.x - 1, y: santa.y)
      elif ch == '>': santa = Position2D(x: santa.x + 1, y: santa.y)
      positions.incl(santa)
  return positions.len()

when isMainModule:
  if paramCount() < 1:
    quit "usage: " & getAppFilename() & " <input file>", 1

  let filename = paramStr(1)
  try:
    echo "result = ", processFile(filename)
  except OSError as e:
    quit "error while processing file `" & filename & "': " & e.msg, 1
